tests/phpunit/tests/html-api/wpHtmlProcessorSemanticRules.php

```diff
@@ -16,6 +16,110 @@ class Tests_HtmlApi_WpHtmlProcessorSemanticRules extends WP_UnitTestCase {
+
+		$this->assertTrue( $p->step(), 'Found no further tags when it should have found the closing DIV' );
+		$this->assertEquals( 'DIV', $p->get_tag(), "Did not skip unexpected BUTTON; stopped at {$p->get_tag()}." );
+		$this->assertTrue( $p->is_tag_closer(), 'Did not find that the terminal DIV tag is a closer.' );
+	}
+
+	/**
+	 * Verifies insertion of a BUTTON element when no existing BUTTON is already in scope.
+	 *
+	 * @ticket 58961
+	 *
+	 * @since 6.4.0
+	 *
+	 * @throws WP_HTML_Unsupported_Exception
+	 */
+	public function test_in_body_button_with_no_button_in_scope() {
+		$p = WP_HTML_Processor::createFragment( '<div><p>Click the button <button one>here</button>!</p></div><button two>not here</button>' );
+
+		$this->assertTrue( $p->next_tag( 'BUTTON' ), 'Could not find expected first button.' );
+		$this->assertTrue( $p->get_attribute( 'one' ), 'Failed to match expected attribute on first button.' );
+		$this->assertSame( array( 'HTML', 'BODY', 'DIV', 'P', 'BUTTON' ), $p->get_breadcrumbs(), 'Failed to produce expected DOM nesting for first button.' );
+
+		$this->assertTrue( $p->next_tag( 'BUTTON' ), 'Could not find expected second button.' );
+		$this->assertTrue( $p->get_attribute( 'two' ), 'Failed to match expected attribute on second button.' );
+		$this->assertSame( array( 'HTML', 'BODY', 'BUTTON' ), $p->get_breadcrumbs(), 'Failed to produce expected DOM nesting for second button.' );
+	}
+
+	/**
+	 * Verifies what when inserting a BUTTON element, when a BUTTON is already in scope,
+	 * that the open button is closed with all other elements inside of it.
+	 *
+	 * @ticket 58961
+	 *
+	 * @since 6.4.0
+	 *
+	 * @throws WP_HTML_Unsupported_Exception
+	 */
+	public function test_in_body_button_with_button_in_scope_as_parent() {
+		$p = WP_HTML_Processor::createFragment( '<div><p>Click the button <button one>almost<button two>here</button>!</p></div><button three>not here</button>' );
+
+		$this->assertTrue( $p->next_tag( 'BUTTON' ), 'Could not find expected first button.' );
+		$this->assertTrue( $p->get_attribute( 'one' ), 'Failed to match expected attribute on first button.' );
+		$this->assertSame( array( 'HTML', 'BODY', 'DIV', 'P', 'BUTTON' ), $p->get_breadcrumbs(), 'Failed to produce expected DOM nesting for first button.' );
+
+		$this->assertTrue( $p->next_tag( 'BUTTON' ), 'Could not find expected second button.' );
+		$this->assertTrue( $p->get_attribute( 'two' ), 'Failed to match expected attribute on second button.' );
+		$this->assertSame( array( 'HTML', 'BODY', 'DIV', 'P', 'BUTTON' ), $p->get_breadcrumbs(), 'Failed to produce expected DOM nesting for second button.' );
+
+		$this->assertTrue( $p->next_tag( 'BUTTON' ), 'Could not find expected third button.' );
+		$this->assertTrue( $p->get_attribute( 'three' ), 'Failed to match expected attribute on third button.' );
+		$this->assertSame( array( 'HTML', 'BODY', 'BUTTON' ), $p->get_breadcrumbs(), 'Failed to produce expected DOM nesting for third button.' );
```

Comment:
[The explanation above one of the test cases](https://github.com/WordPress/wordpress-develop/pull/4955/files#diff-3ca01760e76cd6dca9ea624eb693b645a540cfe995a0636e6f95bef8f3855df9R107-R112) in the next test method is really helpful -- how about adding something like that here as well?

Reply:
I expanded all the test cases in 5ad4fd616e. it's probably a good idea to build these specific rule checks narratively since they are heavily technical and specific, but the assertions are indirect at times.